casper/src/main/scala/coop/rchain/casper/blocks/merger/CasperDagMerger.scala

```diff
@@ -0,0 +1,180 @@
+                                b =>
+                                  fs2.Stream.eval(
+                                    MergingBranchIndex
+                                      .create(b._1._2)
+                                      .map { i =>
+                                        (b._1._1.stateHash, i, b._2)
+                                      }
+                                  )
+                              )
+                            )
+                            .parJoinUnbounded
+                            .compile
+                            .toList
+                            .map(_.sortBy(_._3).map(v => (v._1, v._2)))
+                        )
+
+      // Caching readers for data in base state
+      baseDataReader <- LazyKeyValueCache(
+                         (ch: Blake2b256Hash) =>
+                           baseStateReader
+                             .getDataFromChannelHash(ch)
+                       )
+      baseContReader <- LazyKeyValueCache(
+                         (ch: Blake2b256Hash) =>
+                           baseStateReader
+                             .getContinuationFromChannelHash(ch)
+                       )
+
+      _ <- Log[F].info(
+            s"trying to merge branches size of:  ${branchesIndexed.map(_._2.deploys.map(_.deployLog).size).mkString(";")}"
+          )
+      // detect conflicts across all branches with map reduce
+      toReject <- Stopwatch.time(Log[F].info(_))("CALCULATION DEPLOYS TO REJECT")(
+                   (branchesIndexed.map(_._2), Set.empty[ProcessedDeploy])
+                     .tailRecM[F, Set[ProcessedDeploy]] {
+                       case (_ :: Nil, r) =>
+                         r.asRight[(List[MergingBranchIndex[F]], Set[ProcessedDeploy])].pure[F]
+                       // intermediate result, on each go N branches turn into N/2 (plus leftover option,
+                       // if num of branches is odd)
+                       case (branches, acc) => {
+                         val pairsSize = (branches.size / 2)
+                         val chunks    = branches.sliding(pairsSize, pairsSize).toList
+                         val (pairs, leftover) =
+                           (chunks.head zip chunks(1), chunks.get(3).map(_.head))
+
+                         val workers = pairs.map {
+                           case (main, merging) =>
+                             fs2.Stream.eval(for {
+                               conflictingChans <- ConflictDetectors.findConflics(
+                                                    main,
+                                                    merging,
+                                                    baseDataReader,
+                                                    baseContReader
+                                                  )
+                               rejectedDeploys = merging.calculateRejected(conflictingChans)
+                             } yield (main + merging, rejectedDeploys))
```

Comment:
`main + merging`. What if something in `merging` is conflict? Can you still simply add?

Reply:
I think yes. This sum is a sum of indexes, so the only effect here is that sum index is more "conflicting", Just makes more possible that channel will be conflicting again on the next round of educing. But If channel is conflicting, then it will be removed any way, so thats ok.

Reply:
From your perspective, it makes sense. I am OK with the idea you got here now. 
But I think it is also a core part of consensus (incentive part) like calculating [the order of MergingVertex](https://github.com/rchain/rchain/pull/3238#discussion_r556280427) which we can discuss later.
From the validator perspective, I am more for greedy way.